Add double encryption at S3 gateway.

poornas opened the pull request:
## Description
<!--- Describe your changes in detail -->
This PR implements the API calls needed for s3 gateway encryption - and provides three modes of operation.

1.  Pass-through encryption
2. Single encryption at gateway
3. Double encryption (at gateway and pass-through to backend)

<!--- Provide a general summary of your changes in the Title above -->
If Vault is set up as KMS, s3 gateway will automatically perform single encryption
If MINIO_GW_SSE is set up in addition to Vault KMS, double encryption is performed
when neither KMS or MINIO_GW_SSE is set, do a pass through to backend.

When double encryption is specified, MINIO_GW_SSE can be set to "C" for SSE-C encryption at gateway and backend, "S3" for SSE-S3 encryption at gateway/backend or both to support more than one option.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Facilitate better security by enabling encryption at gateway
## Regression
<!-- Is this PR fixing a regression? (Yes / No) --> No
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Handwritten tests using minio-go SDK against S3 gateway with Vault enabled/disabled and MINIO_GW_SSE set.
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

aead commented on cmd/object-handlers.go:
```diff
@@ -1507,16 +1583,25 @@ func (api objectAPIHandlers) PutObjectPartHandler(w http.ResponseWriter, r *http
 		if crypto.IsEncrypted(li.UserDefined) {
-			if !hasServerSideEncryptionHeader(r.Header) {
-				writeErrorResponse(w, ErrSSEMultipartEncrypted, r.URL)
+			// if !crypto.S3.IsRequested(r.Header) && sseS3 {
```
Remove commented code

aead commented on cmd/object-api-utils.go:
```diff
@@ -302,3 +304,30 @@ type byBucketName []BucketInfo
+// cleanMinioInternalMetadataKeys removes X-Amz-Meta- prefix from minio internal
+// encryption metadata that was sent by minio gateway
+func cleanMinioInternalMetadataKeys(metadata map[string]string) map[string]string {
+	var newMeta = make(map[string]string)
```
Suggestion: preallocate the map. `newMeta := make(map[string]string, len(metadata))`

aead commented on cmd/object-api-utils.go:
```diff
@@ -302,3 +304,30 @@ type byBucketName []BucketInfo
+
+// returns true if object is sse encrypted at the gateway
+func isGatewayEncrypt(metadata map[string]string, headers http.Header) bool {
+	if GlobalGatewaySSE != nil && (hasServerSideEncryptionHeader(headers) || crypto.SSECopy.IsRequested(headers)) {
```
Why is `crypto.SSECopy.IsRequested(headers))` necessary? From my understanding the function returns true if
the object is encrypted. This does/should not depend whether the client sends SSE-C/SSE-C copy headers.

poornas replied:
this was not necessary and has been removed.

aead commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -336,7 +341,12 @@ func (l *s3Objects) GetObject(ctx context.Context, bucket string, key string, st
 // GetObjectInfo reads object info and replies back ObjectInfo
 func (l *s3Objects) GetObjectInfo(ctx context.Context, bucket string, object string, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
-	oi, err := l.Client.StatObject(bucket, object, miniogo.StatObjectOptions{miniogo.GetObjectOptions{ServerSideEncryption: opts.ServerSideEncryption}})
+	var o minio.ObjectOptions
```
Simplify: 
```
statOptions := miniogo.StatObjectOptions{miniogo.GetObjectOptions{
        ServerSideEncryption: opts.ServerSideEncryption,
}}
```

aead commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -313,8 +317,9 @@ func (l *s3Objects) GetObject(ctx context.Context, bucket string, key string, st
 	opts := miniogo.GetObjectOptions{}
-	opts.ServerSideEncryption = o.ServerSideEncryption
-
+	if o.ServerSideEncryption != nil {
```
`nil` check not required - if not set `opts.ServerSideEncryption` is `nil` anyway.

aead commented on cmd/gateway/s3/gateway-s3-sse.go:
```diff
@@ -0,0 +1,629 @@
+
+		var wg = &sync.WaitGroup{}
+		go func(pInfo minio.ObjectInfo) {
+			wg.Add(1)
```
Never call `wg.Add(1)` **inside** goroutine. It's not guaranteed to be called ever.

poornas replied:
removed this.

poornas committed:
Add gateway encryption support for S3 gateway

This PR adds pass-through, single encryption at gateway and double
encryption support (gateway encryption with pass through of SSE
headers to backend).

If KMS is set up (either with Vault as KMS or using
MINIO_SSE_MASTER_KEY),gateway will automatically perform
single encryption. If MINIO_GATEWAY_SSE is set up in addition to
Vault KMS, double encryption is performed.When neither KMS nor
MINIO_GATEWAY_SSE is set, do a pass through to backend.

When double encryption is specified, MINIO_GATEWAY_SSE can be set to
"C" for SSE-C encryption at gateway and backend, "S3" for SSE-S3
encryption at gateway/backend or both to support more than one option.

Fixes #6323, #6696

harshavardhana reviewed (changes requested):
Can you bring the options to GetObjectNInfo in a separate PR which can be merged faster?

kannappanr commented on cmd/gateway/s3/gateway-s3-sse.go:
```diff
@@ -179,8 +179,8 @@ func (l *s3EncObjects) GetObject(ctx context.Context, bucket string, key string,
 	if len(dmeta.Parts) == 0 {
 		// custom gateway encrypted objects uploaded with single PUT operation
 		return l.s3Objects.GetObject(ctx, bucket, getGWContentPath(key), startOffset, length, writer, etag, o)
+		// custom multipart gateway encrypted objects
```
should this comment be moved back to where it was?

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+// key to be sent to the s3 backend for encryption at the backend.
+func deriveClientKey(header http.Header, headerName, bucket, object string) ([32]byte, error) {
+	var key [32]byte
+	clientKey, err := base64.StdEncoding.DecodeString(header.Get(headerName))
```
Pass the key directly: `deriveClientKey(key [32]byte, bucket, object string) [32]byte`

poornas replied:
done

aead commented on cmd/gateway/s3/gateway-s3-metadata.go:
```diff
@@ -0,0 +1,237 @@
+
+// Converts metadata to object info.
+func (m gwMetaV1) ToObjectInfo(bucket, object string) minio.ObjectInfo {
+	filterKeys := append([]string{
```
Reverse order: `append(defaultFilterKeys, "Etag", ...)`

poornas replied:
done

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			if k == GatewaySSEC && crypto.SSEC.IsRequested(r.Header) {
+				derivedKey, err := deriveClientKey(r.Header, crypto.SSECKey, bucket, object)
```
Use `crypto.SSEC.ParseHTTP(r.Header)` and pass the key directly to `deriveClientKey`

poornas replied:
done

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+				}
+				encryption, err = encrypt.NewSSEC(derivedKey[:])
+				if err != nil {
+					return opts, err
```
Probably `logger.Criticalif` - this situations can never happen b/c either the client key was malformed (should have failed before) or the client key is well-formed, so should be the derived key.

poornas replied:
good point - changed this.

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+			// In the case of multipart custom format, the metadata needs to be checked in addition to header to see if it
+			// is SSE-S3 encrypted, primarily because S3 protocol does not require SSE-S3 headers in PutObjectPart calls
+			if k == GatewaySSES3 && (crypto.S3.IsRequested(r.Header) || crypto.S3.IsEncrypted(metadata)) {
+				opts = ObjectOptions{ServerSideEncryption: encrypt.ServerSide(encrypt.NewSSE())}
```
No cast (`encrypt.ServerSide(...)`) necessary.

poornas replied:
fixed.

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			if k == GatewaySSEC && crypto.SSECopy.IsRequested(r.Header) {
+				derivedKey, err := deriveClientKey(r.Header, crypto.SSECopyKey, bucket, object)
```
Same - use `crypto.SSECopy.ParseHTTP(...)` and pass key directly.

poornas replied:
done

kannappanr commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1160,142 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+}
+
+// extract encryption options for pass through to backend in the case of gateway
+func extractEncryptionOption(header http.Header, copySource bool, metadata map[string]string) (opts ObjectOptions, err error) {
```
Can unit tests be added for `extractEncryptionOption`?

poornas replied:
done

kannappanr commented on cmd/gateway-common.go:
```diff
@@ -321,3 +343,34 @@ func ErrorRespToObjectError(err error, params ...string) error {
+// parse gateway sse env variable
+func parseGatewaySSE(s string) ([]string, error) {
+	l := strings.Split(s, ";")
+	gwSlice := make([]string, len(l))
```
Can be  changed to `var gwSlice []string`?, the existing code allocates memory and then appends values to the end.

poornas replied:
done

kannappanr commented on cmd/gateway-common.go:
```diff
@@ -321,3 +343,34 @@ func ErrorRespToObjectError(err error, params ...string) error {
 }
+
+// parse gateway sse env variable
+func parseGatewaySSE(s string) ([]string, error) {
```
Can unit tests be added for `parseGatewaySSE`?

poornas replied:
done

kannappanr commented on cmd/object-api-utils.go:
```diff
@@ -659,3 +657,17 @@ func sealETagFn(key crypto.ObjectKey) SealMD5CurrFn {
+// CleanMinioInternalMetadataKeys removes X-Amz-Meta- prefix from minio internal
+// encryption metadata that was sent by minio gateway
+func CleanMinioInternalMetadataKeys(metadata map[string]string) map[string]string {
+	var newMeta = make(map[string]string, len(metadata))
```
Can this be `var newMeta = make(map[string]string)` and also should we add unit tests for this function?

poornas replied:
since the number of items in the map is known, it is good to provide size hint while allocating.

aead replied:
Completely agree - always pre-allocate if expected size is known and avoid internal resizing.

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -223,9 +224,15 @@ func (g *S3) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, error)
 		Client: clnt,
-	}, nil
+	}
+	if minio.GlobalKMS != nil {
```
The old code could have stayed in place and new code 
```
if minio.GlobalKMS != nil {
encS := s3EncObjects{s}
		go encS.cleanupStaleMultipartUploads(context.Background(), minio.GlobalMultipartCleanupInterval, minio.GlobalMultipartExpiry, minio.GlobalServiceDoneCh)
		return &encS, nil
	}
```
could be above the old code.

```
return &s3Objects{
		Client: clnt,
	}, nil
```

poornas replied:
s3EncObjects embeds s3Objects struct.hence the need for a temp variable.

kannappanr commented on cmd/gateway-common.go:
```diff
@@ -321,3 +343,34 @@ func ErrorRespToObjectError(err error, params ...string) error {
+
+// handle gateway env vars
+func handleGatewayEnvVars() {
+	gwsse, ok := os.LookupEnv("MINIO_GW_SSE")
```
Can this be named `MINIO_GATEWAY_SSE` instead of `MINIO_GW_SSE`. We already have an environment variable named `MINIO_GATEWAY_DEPLOYMENT_ID`. `MINIO_GATEWAY_SSE` might be more consistent with it.

poornas replied:
yes, we can rename it to MINIO_GATEWAY_SSE

kannappanr commented on cmd/gateway-main.go:
```diff
@@ -215,35 +218,39 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 
 		// Override any values from ENVs.
 		srvCfg.loadFromEnvs()
 
```
`handleGatewayEnvVars()` should be moved here or after this `if-else` block.

poornas committed:
Support passthrough encryption at S3 gateway

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1159,119 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+// The customer key in the header is used by the gateway for encryption in the case of
+// s3 gateway double encryption. A new client key is derived from the customer provided
+// key to be sent to the s3 backend for encryption at the backend.
+func deriveClientKey(clientKey [32]byte, bucket, object string) ([32]byte, error) {
```
The error in the function signature is not used. It should be removed.

poornas replied:
thanks,fixed this

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1159,119 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+
+	if copySource {
+		if crypto.SSECopy.IsRequested(header) {
+			clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECopyKey))
```
That parsing is already implemented in [`SSECopy.ParseHTTP`](https://github.com/minio/minio/blob/master/cmd/crypto/header.go#L199)

poornas replied:
fixed this

aead commented on cmd/encryption-v1.go:
```diff
@@ -1157,3 +1159,119 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
+	}
+
+	if crypto.SSEC.IsRequested(header) {
+		clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECKey))
```
Same, already implemented in [`SSEC.ParseHTTP`](https://github.com/minio/minio/blob/master/cmd/crypto/header.go#L174)

poornas replied:
done

aead reviewed (commented):
Tested the functionality - everything seems to work, awesome job @poornas :tada: 
LGTM - only minor suggestions for some code.

poornas committed:
fix review comments

harshavardhana reviewed (dismissed):
LGTM :cake:

poornas committed:
fix for review comments
 ensure gateway sse mode before passing sse options to backend
 fix for travis test failure

kannappanr reviewed (approved):
Tested and LGTM.
